vpn/client.go

```diff
@@ -0,0 +1,180 @@
+package vpn
+
+import (
+	"context"
+	"net/http"
+	"net/netip"
+	"net/url"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"tailscale.com/net/dns"
+	"tailscale.com/wgengine/router"
+
+	"github.com/tailscale/wireguard-go/tun"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/workspacesdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/quartz"
+)
+
+type Conn interface {
+	CurrentWorkspaceState() *proto.WorkspaceUpdate
+	Close() error
+}
+
+type vpnConn struct {
+	*tailnet.Conn
+
+	cancelFn    func()
+	controller  *tailnet.Controller
+	updatesCtrl *tailnet.TunnelAllWorkspaceUpdatesController
+}
+
+func (c vpnConn) CurrentWorkspaceState() *proto.WorkspaceUpdate {
+	return c.updatesCtrl.CurrentState()
+}
+
+func (c vpnConn) Close() error {
```

Comment:
This is not bugged, but it feels a bit slippery to call `Close()` on a struct by value. It so happens that all the internal state of `vpnConn` is itself reference state, but I could easily imagine someone coming in and adding some non-referential state to `vpnConn`, and then passing by value will cause bugs.  Safer to switch to methods on the pointer, since "Close" is, conceptually, a modification.